rust/src/container.rs

```diff
@@ -659,3 +800,413 @@ pub(crate) fn deploy_from_self_entrypoint(args: Vec<String>) -> CxxResult<()> {
+
+                    // Create xattr tuples as (name, value) pairs where both are byte arrays
+                    // The function expects the name to be "user.component\0" and value to be "test-component"
+                    let xattr_data = vec![(&b"user.component\0"[..], &b"test-component"[..])];
+
+                    // Convert to glib::Variant
+                    glib::Variant::from(xattr_data)
+                } else {
+                    // Return empty xattr array for other files
+                    let empty_xattrs: Vec<(&[u8], &[u8])> = vec![];
+                    glib::Variant::from(empty_xattrs)
+                }
+            };
+
+        modifier.set_xattr_callback(xattr_callback);
+
+        // Write directory to mutable tree using OSTree API
+        let work_dir_fd =
+            cap_std::fs::Dir::open_ambient_dir(&work_dir, cap_std::ambient_authority())?;
+        repo.write_dfd_to_mtree(
+            work_dir_fd.as_raw_fd(),
+            ".",
+            &mtree,
+            Some(&modifier),
+            gio::Cancellable::NONE,
+        )?;
+
+        // Write mutable tree to get root directory
+        let root = repo.write_mtree(&mtree, gio::Cancellable::NONE)?;
+
+        // Convert gio::File to ostree::RepoFile for write_commit
+        let root_repo_file = root
+            .downcast::<ostree::RepoFile>()
+            .map_err(|_| "Failed to downcast root to RepoFile")?;
+
+        // Create metadata for the commit
+        let metadata = glib::VariantDict::new(None);
+        metadata.insert("test", &"xattr-test");
+        let metadata_variant = metadata.to_variant();
+
+        // Write commit using OSTree API
+        let commit_checksum = repo.write_commit(
+            None,                           // parent
+            Some("Test commit"),            // subject
+            Some("Test commit with xattr"), // body
+            Some(&metadata_variant),        // metadata
+            &root_repo_file,                // root
+            gio::Cancellable::NONE,
+        )?;
+
+        // Read back the commit to get our RepoFile
+        let (commit_root, _) = repo.read_commit(&commit_checksum, gio::Cancellable::NONE)?;
+        let test_file = commit_root.child("test-file");
+        let repo_file = test_file
+            .downcast::<ostree::RepoFile>()
+            .map_err(|_| "Failed to downcast test-file to RepoFile")?;
+
+        // Test our function with the file that should have the xattr
+        let result = get_user_component_xattr(&repo_file)?;
+
+        // Verify the function found the xattr and contains our test component
+        assert!(result.is_some());
+        let result_str = result.unwrap();
+        assert!(result_str.contains("test-component"));
+
+        // Test with a file that should have no component xattr (root directory)
+        let root_repo_file = commit_root
+            .downcast::<ostree::RepoFile>()
+            .map_err(|_| "Failed to downcast root to RepoFile")?;
```

Comment:
This should never fail, and especially in a unit test it's OK to `unwrap()` for me FWIW.